File: go/101SymmetricTree.go
package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}

	var isMirror func(*TreeNode, *TreeNode) bool
	isMirror = func(node1, node2 *TreeNode) bool {
		if node1 == nil && node2 == nil {
			return true
		}

		if (node1 == nil && node2 != nil) || (node1 != nil && node2 == nil) || (node1.Val != node2.Val) {
			return false
		}

		return isMirror(node1.Left, node2.Right) && isMirror(node1.Right, node2.Left)
	}

	return isMirror(root.Left, root.Right)
}

func main() {
	output := isSymmetric(
		&TreeNode{
			Val: 1,
			Left: &TreeNode{
				Val: 2,
				Left: &TreeNode{
					Val:   3,
					Left:  nil,
					Right: nil},
				Right: &TreeNode{
					Val:   4,
					Left:  nil,
					Right: nil},
			},
			Right: &TreeNode{
				Val: 2,
				Left: &TreeNode{
					Val:   4,
					Left:  nil,
					Right: nil},
				Right: &TreeNode{
					Val:   3,
					Left:  nil,
					Right: nil},
			},
		})
	output1 := isSymmetric(
		&TreeNode{
			Val: 1,
			Left: &TreeNode{
				Val:  2,
				Left: nil,
				Right: &TreeNode{
					Val:   3,
					Left:  nil,
					Right: nil},
			},
			Right: &TreeNode{
				Val:  2,
				Left: nil,
				Right: &TreeNode{
					Val:   3,
					Left:  nil,
					Right: nil},
			},
		})

	fmt.Println("\n******** output *********\n", output, output1)
}

File: cpp/101SymmetricTree.cpp
#include <algorithm>
#include <cstdint>
#include <cstdlib>
#include <functional>
#include <iostream>
#include <iterator>
#include <limits>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

template<class Container>
void printContainer(const Container& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[";
        }
        cout << *it;
        if (it == prev(container.end())) {
            cout << "]" << endl;
        } else {
            cout << ", ";
        }
    }
}

template<class T>
void print2DimVec(const vector<vector<T>>& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[" << endl;
        }
        printContainer(*it);
        if (it == prev(container.end())) {
            cout << "]" << endl;
        }
    }
}

template<class T>
void printMap(const T& container) {
//    return;

    cout << "size " << container.size() << endl;
    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }
    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[ ";
        }

        cout << "[" << it->first << ", " << it->second << "]";

        if (next(it) == container.end()) {
            cout << " ]" << endl;
        } else {
            cout << ", ";
        }
    }
}


/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

void printList(const ListNode* node) {
    while (node != nullptr) {
        cout << node->val;

        node = node->next;
        
        if (node != nullptr) {
            cout << "->";
        } else {
            cout << endl;
        }
    }
}

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {
    }
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {
    }
    TreeNode(int x, TreeNode *left, TreeNode *right)
        : val(x), left(left), right(right) {
    }
};

class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        function<bool(TreeNode*, TreeNode*)> isMirror = [&](TreeNode* node1,
                                                            TreeNode* node2) {
            if (node1 == nullptr && node2 == nullptr) {
                return true;
            }

            if ((node1 == nullptr && node2 != nullptr) ||
                (node1 != nullptr && node2 == nullptr)) {
                return false;
            }

            if (node1->val != node2->val) {
                return false;
            }

            return isMirror(node1->left, node2->right) &&
                   isMirror(node1->right, node2->left);
        };

        if (root == nullptr) {
            return true;
        }

        return isMirror(root->left, root->right);
    }
};

int main() {
    Solution s;

    TreeNode node021(3);
    TreeNode node022(4);
    TreeNode node023(4);
    TreeNode node024(3);
    TreeNode node011(2, &node021, &node022);
    TreeNode node012(2, &node023, &node024);
    TreeNode node001(1, &node011, &node012);
    auto output = s.isSymmetric(&node001);

    TreeNode node122(3);
    TreeNode node124(3);
    TreeNode node111(2, nullptr, &node122);
    TreeNode node112(2, nullptr, &node124);
    TreeNode node101(1, &node111, &node112);
    auto output1 = s.isSymmetric(&node101);

    cout << "*************output*************" << endl;
    cout << output << endl;
    cout << output1 << endl;

    return 0;
}
